rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -213,21 +212,22 @@ private boolean allowByMinerNotBanned(Peer sender, Message message) {
     private boolean allowByMessageUniqueness(Peer sender, Message message) {
         Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
 
-        boolean contains = receivedMessages.contains(encodedMessage);
+        Set<Keccak256> receivedMessages = this.receivedPeerMessages.computeIfAbsent(sender, k -> Collections.synchronizedSet(new HashSet<>()));
 
-        if (!contains) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
-                if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
-                    this.receivedMessages.clear();
-                }
-                this.receivedMessages.add(encodedMessage);
+        if (receivedMessages.contains(encodedMessage)) {
+            reportEventToPeerScoring(sender, EventType.REPEATED_MESSAGE, "Received repeated message on {}, not added to the queue");
+            return false;
+        }
+
+        if (Stream.of(MessageType.BLOCK_MESSAGE, MessageType.TRANSACTIONS).anyMatch(t -> message.getMessageType() == t)) {
```

Comment:
I would keep the `if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS)` instead of the stream. Is more efficient and simply to read.